yang/frr-bfdd.yang

```diff
@@ -242,6 +242,40 @@ module frr-bfdd {
     }
   }
 
+  grouping bfd-monitoring {
+    description
+      "BFD monitoring template for protocol integration.";
+
+    leaf enable {
```

Comment:
it seems that when bfd-monitoring only contains "enable true", it is single-hop. It is not clear enough.

`source`, `multi-hop` and `profile` cannot be set together. You need have `choice`

```
choice type {
  case single-hop {
    leaf single-hop {
      type boolean;
    }
    leaf source {
      (...)
    }
  }
  case multi-hop {
    leaf multi-hop {
      type boolean;
    }
  }
  case profile {
    leaf profile {
      description
        "BFD pre configured profile.";
      type frr-bfdd:profile-ref;
    }
  }
```
In that way, it will be easy to add the `auto-hop` option to automatically select single-hop or multi-hop

Reply:
`enable` only toggles BFD session status (`true` or `false`). The `multi-hop` toggle is a leaf one item below (defaults to `false`), you can set it to `true` if you want.

You can always use `profile` if you want.

I'm very confused by this `choice type` you are proposing.

Reply:
> You can always use `profile` if you want.

OK. You can also have a `source` with `multi-hop`. No need for `choice` then.

However, `enable` is confusing. It is the administrative state of the BFD session but there is no CLI option for that. You never do `enable false` but you remove the BFD monitored route in the CLI.

I think the following is a better option.

```
        leaf hop-mode {
          description
            "Hop mode";
          default "single-hop";
          type enumeration {
              enum single-hop {
              value 0;
              description
                "single-hop BFD session";
            }
            enum multi-hop {
            value 1;
              description
                "multi-hop BFD session";
            }
          }
        }
```

In the future:

```
            enum auto {
              value 2;
              description
                "Select automatically the BFD mode: single or multi-hop";
            }
```

Reply:
> However, enable is confusing. It is the administrative state of the BFD session but there is no CLI option for that. You never do enable false but you remove the BFD monitored route in the CLI.

Unfortunately the `enable` toggle doesn't exist for static route CLI, but it will for other daemons integration. For an example: `bgpd`. See discussion here: https://github.com/FRRouting/frr/pull/11951#issuecomment-1330776764

I'm implementing it here because it will eventually be used and the northbound handling is also implemented for `staticd` (even though there is no CLI).

> I think the following is a better option.
> ```yang
> leaf hop-mode {
> ```

Sorry I'm not implementing auto hop, but I can help you implement/review it later if you want.

Reply:
I did not find where you set enable in bgp CLI:
You do not do `neighbor 172.16.1.1 bfd enabled` but just `neighbor 172.16.1.1 bfd`. If you want to disable BFD, you just remove the command.

Reply:
@louis-6wind it doesn't exist yet. The new `grouping bfd-monitoring` will be a template that protocols should use to integrate with `bfdd`. In BGP's case `neighbor A.B.C.D bfd` will toggle the `enable` leaf, today it doesn't even use YANG/northbound.

Same will apply to `ospfd`, `ospf6d`, `pimd`, `isisd` and `fabricd`:
- `ip ospf bfd`
- `ipv6 ospf6 bfd`
- `isis bfd`
- `ip pim bfd`

This is needed to satisfy the condition ( https://github.com/FRRouting/frr/pull/11951#discussion_r975526265 ) of allowing protocols to save BFD configuration but not have it enabled. After these changes these protocols should be able to configure `ip ospf bfd profile FOO`, for an example, and save configuration without having BFD enabled.

Reply:
> enable is not needed unless you have a CLI enable option, which is not the case.

Sure. I'll remove it because it doesn't make sense for this particular case, but when the other daemons work start it will come back anyway.

I'll push the changes soon.

> May be moving container bfd-monitoring under grouping bfd-monitoring would be better

Makes sense, I didn't think of this before.

Reply:
@louis-6wind I removed the `enable` leaf as requested.

I could not move the `bfd-monitoring` container to the grouping and it didn't make sense because otherwise we'd introduce an extra container level. The outer container in `frr-staticd.yang` is required because I introduced some conditionals to allow the existence of BFD: BFD monitoring can only be used with IPv4/IPv6 gateways / routes. I think other protocols might want to add conditionals as well, so let the user protocol decide how the YANG tree looks like.

Reply:
OK. Thank you for the change